### PR TITLE
feat(compile): support discovering modules for more dynamic arguments

### LINKED ISSUE #21390: suggestion: option to ignore missing JSDocs in remote modules for `deno doc --lint`
`deno doc --lint` also scans remote modules for missing JSDocs. I don't know if this behaviour is expected. If it is, what are some ways to make the linter happy without control of the remote module (able to add JSDocs)?

The linter is failing for https://github.com/denoland/deno_kv_oauth/pull/280, and I don't know how to proceed.

### LINKED ISSUE #21403: bug: `deno doc --lint` reports missing return type when missing `void` or `Promise<void>`
To reproduce, run `deno doc --lint https://deno.land/std/archive/untar.ts`. You'll get the following:
```
...
Missing return type.
    at https://deno.land/std@0.208.0/archive/untar.ts:144:3
...
```
[The method returns `Promise<void>`](https://deno.land/std@0.208.0/archive/untar.ts?source=#L144).

This also appears to be the case for `void`. E.g. run `deno doc --lint https://deno.land/std/fmt/colors.ts`, and you'll get the following:
```
...
Missing return type.
    at file:///Users/asher/GitHub/deno_std/fmt/colors.ts:78:1
...
```

As `void` and `Promise<void>` are seen as needless boilerplate, I'd expect this error not to come up.

Or is this intentional?

### LINKED ISSUE #21405: bug: `deno doc --lint` reports error when class constructor is missing JSDoc comment
To reproduce, run `deno doc --lint https://deno.land/std/async/deadline.ts`. The following will be printed:
```
Missing JSDoc comment.
    at file:///Users/asher/GitHub/deno_std/async/deadline.ts:14:3
```
The line is the constructor for the `DeadlineError` class. Is this intentional?

### PR DESCRIPTION
**Pull request by dsherret**
This PR causes Deno to include more files in the graph based on how a template literal looks that's provided to a dynamic import:

```ts
const file = await import(`./dir/${expr}`);
```

In this case, it will search the `dir` directory and descendant directories for any .js/jsx/etc modules and include them in the graph.

To opt out of this behaviour, move the template literal to a separate line:

```ts
const specifier = `./dir/${expr}`
const file = await import(specifier);
```

Also, closes (due to deno_doc upgrade):

Closes #21405
Closes #21403
Closes #21390

### CODE REVIEW

File: cli/util/import_map.rs [hunk 1]
```diff
@@ -75,7 +76,20 @@ impl<'a> ImportMapUnfurler<'a> {
...
-      analyze_specifier(&dep.specifier, &dep.specifier_range);
+      match dep {
+        DependencyDescriptor::Static(dep) => {
+          analyze_specifier(&dep.specifier, &dep.specifier_range);
+        }
+        DependencyDescriptor::Dynamic(dep) => match &dep.argument {
+          deno_graph::DynamicArgument::String(arg) => {
+            analyze_specifier(arg, &dep.argument_range);
+          }
+          deno_graph::DynamicArgument::Template(_) => todo!(),
```
Comment (dsherret) [inline cli/util/import_map.rs:87]: Need to do this.

File: cli/util/import_map.rs [hunk 2]
```diff
@@ -59,23 +63,49 @@ impl<'a> ImportMapUnfurler<'a> {
...
+            log::warn!(
+              "{} Dynamic import was not analyzable and won't use the import map once published.\n    at {}",
+              crate::colors::yellow("Warning"),
+              format_range_with_colors(&deno_graph::Range {
+                specifier: url.clone(),
+                start: dep.range.start.clone(),
+                end: dep.range.end.clone(),
+              })
+            );
+          }
```
Comment (bartlomieju) [inline cli/util/import_map.rs:106]: Could you add a test that exercised this warning?

Comment (dsherret) [reply cli/util/import_map.rs:106]: I wanted to, but we don't have any integration tests for publishing at the moment.

**Comment by iuioiua** [comment]
Awesome. The `deno_doc` upgrade is going to help us document the Standard Library better. Thanks for this!

**Comment by dsherret** [comment]
@iuioiua and thank you for trying out the `--lint` flag and pointing out issues!

**Pull request merged**